File: sdk/.net/ardvro.sdk.net/component/kenviro/controller/SelfOrganizingMapsController.cs
namespace ardvro.sdk.net.component.kenviro.controller;

using ardvro.core.ext.connection;
using ardvro.core.lib.common;

public class SelfOrganizingMapsController
{
    private IConnection _connection;
    public SelfOrganizingMapsController(IConnection connection)
    {
        _connection = connection;
    }

    public void Learning(double[] inputs, double[][] weights, int neuronCounts, double learningRate, double learningRadius, double errorTarget, int epoch, Action<Object> callback)
    {
        Flexible args = new Flexible();
        args.Add("Inputs", inputs);
        args.Add("Weights", weights);

        Flexible setting = new Flexible();
        setting.Add("LearningRate", learningRate);
        setting.Add("LearningRadius", learningRadius);
        setting.Add("ErrorTarget", errorTarget);
        setting.Add("Epoch", epoch);
        setting.Add("InputCount", inputs.Length);
        setting.Add("NeuronCount", neuronCounts);

        args.Add("Setting", setting);

        _connection.Submit(args, "ardvro/component/kenviro/SelfOrganizingMaps/Learning", callback);
    }

    public void Compute(double[][] inputs, int neuronCounts, double learningRate, double learningRadius, double errorTarget, int epoch, Action<object> callback)
    {
        Flexible args = new Flexible();
        args.Add("Inputs", inputs);

        Flexible setting = new Flexible();
        setting.Add("LearningRate", learningRate);
        setting.Add("LearningRadius", learningRadius);
        setting.Add("ErrorTarget", errorTarget);
        setting.Add("Epoch", epoch);
        setting.Add("InputCount", inputs[0].Length);
        setting.Add("NeuronCount", neuronCounts);

        args.Add("Setting", setting);

        _connection.Submit(args, "ardvro/component/kenviro/SelfOrganizingMaps/Learning", callback);
    }

}

File: sample/kenviro/controller/SelfOrganizingMapsController.js
var SelfOrganizingMapsController = function SelfOrganizingMapsController(cfg)
{
    let ctrl = {};

    ctrl.Learning = function (inputs, neuronCounts, learningRate, learningRadius, errorTarget, epoch, callback)
    {
        let args = {
            Inputs: inputs,
            Setting: {
                LearningRate: learningRate,
                LearningRadius: learningRadius,
                ErrorTarget: errorTarget,
                Epoch: epoch,
                InputCount: inputs[0].length,
                NeuronCount: neuronCounts,
            }
        };
        cfg.Connector.Submit(args, "ardvro/component/kenviro/SelfOrganizingMaps/Learning", function (box)
        {
            callback != null ? callback(box == null ? null : box.Data) : null;
        });
    };

    ctrl.Compute = function (inputs, weights, neuronCounts, learningRate, learningRadius, errorTarget, epoch, callback)
    {
        let args = {
            Inputs: inputs,
            Weights: weights,
            Setting: {
                LearningRate: learningRate,
                LearningRadius: learningRadius,
                ErrorTarget: errorTarget,
                Epoch: epoch,
                InputCount: inputs[0].length,
                NeuronCount: neuronCounts,
            }
        };
        cfg.Connector.Submit(args, "ardvro/component/kenviro/SelfOrganizingMaps/Compute", function (box)
        {
            callback != null ? callback(box == null ? null : box.Data) : null;
        });
    };


    return ctrl;
}